src/main.rs

```diff
@@ -160,3 +157,78 @@ fn pass_to_sub(command: String) -> Result<(), MainError> {
         None => MainError::Pn(PnError::UnexpectedTermination { command }),
     })
 }
+
+fn read_package_manifest(manifest_path: &Path) -> Result<NodeManifest, MainError> {
+    manifest_path
+        .pipe(File::open)
+        .map_err(|err| match err.kind() {
+            ErrorKind::NotFound => MainError::Pn(PnError::NoPkgManifest {
+                file: manifest_path.to_path_buf(),
+            }),
+            _ => MainError::from_dyn(err),
+        })?
+        .pipe(serde_json::de::from_reader::<_, NodeManifest>)
+        .map_err(|err| {
+            MainError::Pn(PnError::ParseJSONError {
+                file: manifest_path.to_path_buf(),
+                message: err.to_string(),
+            })
+        })
+}
+
+#[test]
+fn test_read_package_manifest_ok() {
+    use serde_json::to_string_pretty;
+    use std::fs;
+    use tempfile::tempdir;
+
+    let temp_dir = tempdir().unwrap();
+    let package_json_path = temp_dir.path().join("package.json");
+    fs::write(
+        &package_json_path,
+        r#"{"scripts": {"test": "echo hello world"}}"#,
+    )
+    .unwrap();
+
+    let package_manifest = read_package_manifest(&package_json_path).unwrap();
+
+    let received = serde_json::to_string_pretty(&package_manifest).unwrap();
+    let expected = serde_json::json!({
+        "scripts": {
+            "test": "echo hello world"
+        }
+    })
+    .pipe_ref(to_string_pretty)
+    .unwrap();
+
+    assert_eq!(received, expected);
+}
+
+#[test]
+fn test_read_package_manifest_error() {
+    use std::fs;
+    use tempfile::tempdir;
+
+    let temp_dir = tempdir().unwrap();
+    let package_json_path = temp_dir.path().join("package.json");
+    fs::write(
+        &package_json_path,
+        r#"{"scripts": {"test": "echo hello world",}}"#,
+    )
+    .unwrap();
+
+    let read_package_manifest_error_display = format!(
+        "{:?}",
+        read_package_manifest(&package_json_path).unwrap_err()
+    );
+    eprintln!("Read package manifest error:\n{read_package_manifest_error_display}\n");
+
+    let expected_display = format!(
+        "{:?}",
+        MainError::Pn(PnError::ParseJSONError {
+            file: package_json_path,
+            message: String::from("trailing comma at line 1 column 41")
+        })
+    );
+    assert_eq!(read_package_manifest_error_display, expected_display);
```

Comment:
The variable reads `_display` but the formatter is `{:?}` (debug). I am confused. What are you trying to do here? Are you just trying to compare 2 error objects without deriving `PatrialEq`? Or are you just trying to compare 2 error messages?

If it's the latter, then you should just call `.to_string()` for `read_package_manifest_error_display` and `format!("... {package_json_path:?} ...")` for `expected_display`. You should also rename these variables to `received_message` and `expected_message` respectively.

If, however, what you are trying to do is to test if the returning error is of `ParseJsonError`, just write `assert!(matches!(received_error, PnError::ParseJsonError { .. }))` (remember to log the error before doing the assertion).